### PR TITLE
Fix guessing logic when a dot is present

### PR DESCRIPTION
**Pull request by asilvr**
### Description

This is a simplified version of https://github.com/duckduckgo/Android/pull/1530, focusing on improved handling when a `.` character is present in the guess. It also addresses an issue where URLs with multiple path segments and a MIME type was guessing the first segment of the path as the filename, as opposed to the last.

### Steps to test this PR

Navigate to any of the following URLs and observe the filename they're downloaded as

| URL      |  5.101.0 | This branch |
| ----------- | ----------- | ----------- | 
| https://d1x9zcshh2aa46.cloudfront.net/abc.def/example.pdf  | abc.def  | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/abc.123/example.pdf | abc.123  | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/123.abc/example.pdf | 123.abc  | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/123.456/example.pdf | 123.456 | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/abc.pdf/example.pdf  | abc.pdf | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/pdfs/example.pdf | pdfs.pdf | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/pdfs/example | pdfs.pdf | example.pdf |

_Note_: All of the above URLs have a MIME type of `application/pdf` and no content disposition.

### CODE REVIEW

**Comment by asilvr** [comment]
@cmonfortep This PR is a reduced version of #1530 and is ready for a review (the tests I added pass here as well)

**Comment by cmonfortep** [comment]
@asilvr could it be that PR description needs to be updated? I think all the MIME types references apply to the previous PR and not to this reduced version. Could you double check? Thanks.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [inline app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:86]: not sure if this is necessary, can you double check?

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [inline app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:89]: there's a warning because mimeTypes is always non-nullable.

You can use type inference here, and remove the type.
`val mimeType = "image/jpeg"`

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [inline app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:103]: same here.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [inline app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:100]: ditto

**Review by cmonfortep — APPROVED** [review]
👍  thanks, all good. Just left some minor comments in the unit tests.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:86]: From what I gather, `@SdkSuppress(minSdkVersion = 22)` is used in the [existing MIME type tests](https://github.com/duckduckgo/Android/blob/784c87beef30a351436f8a5300950b6f7081c77e/app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt#L167-L195) because in SDK version 21 and below, `image/jpeg` defaults to a `.jpeg` file extension, and SDK 22 and above defaults to a `.jpg` file extension (dropped the `e`). So using it here just ensures that the test doesn't fail if/when we execute it against a lower SDK target.

In the case of this test case, I _think_ it's required to satisfy the assertion, but I will double-check.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:89]: Good catch. Will fix.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:103]: Good catch. Will fix.

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:100]: Since this test contains a guess with an extension which satisfies the MIME type, I believe that any SDK version will recognize this (versus the other test where the extension was ambiguous). I'll remove this and see if it still passes!

**Comment by asilvr** [comment]
> @asilvr could it be that PR description needs to be updated? I think all the MIME types references apply to the previous PR and not to this reduced version. Could you double check? Thanks.

@cmonfortep In reference to this:

> It also addresses an issue where URLs with multiple path segments and a MIME type was guessing the first segment of the path as the filename, as opposed to the last.

IMO, this is relevant to the PR. While this PR itself contains no source code changes related to MIME types, URLs which use MIME types are beneficiaries of the contribution. Specifically, these two URLs from the description get incorrect filename guesses today due to the previous logic:

| URL      |  5.101.0 | This branch |
| ----------- | ----------- | ----------- | 
| https://d1x9zcshh2aa46.cloudfront.net/pdfs/example.pdf | pdfs.pdf | example.pdf |
| https://d1x9zcshh2aa46.cloudfront.net/pdfs/example | pdfs.pdf | example.pdf |

This is because they have multiple path segments, and there was no prior test coverage for responses containing multiple path segments and a MIME type (nearly all the tests don't provide a MIME type, which is totally OK but led to a gap in coverage). I believe the added test cases which specify a MIME type demonstrates the change, and lead to a more logical filename based on the response.

Hopefully that makes sense!

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:100]: Makes sense 👍

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (cmonfortep) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt:86]: Thanks for the explanation, I was not aware of that 👍

**Comment by cmonfortep** [comment]
Thanks for the insights @asilvr. I will wait until we remove those warnings to merge the PR. This contribution is going to improve the user experience, and excited about it 🙌

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt]: Removed this and it passed 🎉

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt]: Fixed this and it passed 🎉

File: app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt
Comment (asilvr) [reply app/src/androidTest/java/com/duckduckgo/app/browser/downloader/UriUtilsFilenameExtractorTest.kt]: Fixed this and it passed 🎉

**Comment by asilvr** [comment]
Should be good to go @cmonfortep! Sorry for the delay, I've been traveling for US holidays. 

Appreciate you helping facilitate this review+merge, and looking forward this user experience improvement. It was a pleasure to contribute!

**Pull request merged**
